### PR TITLE
HBASE-26271: Cleanup the broken store files under data directory

### PR DESCRIPTION
**Pull request by BukrosSzabolcs**
Add new chore to delete lefotver directly stored hfiles
Allow DirectStoreCompactor to list current compaction target paths for
easier validation

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   8m 23s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ HBASE-24749 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 43s |  HBASE-24749 passed  |
| +1 :green_heart: |  compile  |   3m 38s |  HBASE-24749 passed  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  HBASE-24749 passed  |
| +1 :green_heart: |  spotbugs  |   2m 16s |  HBASE-24749 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 32s |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 32s |  the patch passed  |
| -0 :warning: |  checkstyle  |   1m 13s |  hbase-server: The patch generated 15 new + 61 unchanged - 0 fixed = 76 total (was 61)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  22m 25s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1 3.3.0.  |
| -1 :x: |  spotbugs  |   3m  3s |  hbase-server generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 14s |  The patch does not generate ASF License warnings.  |
|  |   |  65m  4s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hbase-server |
|  |  Dead store to fsStoreFiles in org.apache.hadoop.hbase.regionserver.DirectHFileCleaner.chore()  At DirectHFileCleaner.java:org.apache.hadoop.hbase.regionserver.DirectHFileCleaner.chore()  At DirectHFileCleaner.java:[line 91] |
|  |  Random object created and used only once in org.apache.hadoop.hbase.regionserver.HRegionServer.initializeThreads()  At HRegionServer.java:only once in org.apache.hadoop.hbase.regionserver.HRegionServer.initializeThreads()  At HRegionServer.java:[line 2234] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3707 |
| JIRA Issue | HBASE-26271 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti checkstyle compile |
| uname | Linux a3dda2b2f819 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-24749 / 78999ab725 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/artifact/yetus-general-check/output/diff-checkstyle-hbase-server.txt |
| spotbugs | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/artifact/yetus-general-check/output/new-spotbugs-hbase-server.html |
| Max. process+thread count | 86 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   8m 14s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  5s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ HBASE-24749 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 24s |  HBASE-24749 passed  |
| +1 :green_heart: |  compile  |   1m 30s |  HBASE-24749 passed  |
| +1 :green_heart: |  shadedjars  |   9m 13s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 47s |  HBASE-24749 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m  5s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 24s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 24s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   9m  2s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 43s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 215m  8s |  hbase-server in the patch passed.  |
|  |   | 258m 25s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3707 |
| JIRA Issue | HBASE-26271 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux d316e1531da6 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-24749 / 78999ab725 |
| Default Java | AdoptOpenJDK-11.0.10+9 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/testReport/ |
| Max. process+thread count | 3353 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   9m 23s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  4s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ HBASE-24749 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 29s |  HBASE-24749 passed  |
| +1 :green_heart: |  compile  |   1m  9s |  HBASE-24749 passed  |
| +1 :green_heart: |  shadedjars  |   9m 15s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 39s |  HBASE-24749 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 12s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  7s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  7s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   9m 50s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 42s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 234m 44s |  hbase-server in the patch passed.  |
|  |   | 277m 38s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3707 |
| JIRA Issue | HBASE-26271 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 6cbadd9894fa 4.15.0-143-generic #147-Ubuntu SMP Wed Apr 14 16:10:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-24749 / 78999ab725 |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/testReport/ |
| Max. process+thread count | 3272 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3707/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by Apache9 — COMMENTED** [review]
We are working on the HBASE-26067 branch now as it is more general, which makes it possible to use different store file tracking way along with stripe compaction and date tiered compaction.

I think most of code can be reused, just need a rebase.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DirectHFileCleaner.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This Chore, every time it runs, will clear the unsused  direct HFiles in the data
+ * folder that are deletable for each HFile cleaner in the chain.
+ */
+@InterfaceAudience.Private public class DirectHFileCleaner extends ScheduledChore {
```
[the remark this answers is not in the document]
Comment (wchevreuil) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DirectHFileCleaner.java:43]: Talked to @BukrosSzabolcs internally, there was a bit of confusion on which branch has the actual "renameless" solution. That naming was following the pattern of HBASE-24749 branch. He will be submitting a PR compatible with HBASE-26067 soon.

**Pull request closed without merging**